compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala

```diff
@@ -950,9 +951,34 @@ class ClassfileParser(
           if (tastyBytes.nonEmpty) {
             val reader = new TastyReader(bytes, 0, 16)
             val expectedUUID = new UUID(reader.readUncompressedLong(), reader.readUncompressedLong())
-            val tastyUUID = new TastyHeaderUnpickler(tastyBytes).readHeader()
+            val tastyHeader = new TastyHeaderUnpickler(tastyBytes).readFullHeader()
+            val fileTastyVersion = TastyVersion(tastyHeader.majorVersion, tastyHeader.minorVersion, tastyHeader.experimentalVersion)
+            val tastyUUID = tastyHeader.uuid
             if (expectedUUID != tastyUUID)
               report.warning(s"$classfile is out of sync with its TASTy file. Loaded TASTy file. Try cleaning the project to fix this issue", NoSourcePosition)
+
+            val tastyFilePath = classfile.path.stripSuffix(".class") + ".tasty"
+
+            def reportWrongTasty(reason: String, highestAllowed: TastyVersion) =
+              report.error(s"""The class ${classRoot.symbol.showFullName} cannot be loaded from file ${tastyFilePath} because $reason:
+                              |highest allowed: ${highestAllowed.show}
+                              |found:           ${fileTastyVersion.show}
+              """.stripMargin)
+
+            val isTastyReadable = fileTastyVersion.isCompatibleWith(TastyVersion.compilerVersion)
+            if !isTastyReadable then
+              reportWrongTasty("its TASTy format cannot be read by the compiler", TastyVersion.compilerVersion)
+            else
+              def isStdlibClass(cls: ClassDenotation): Boolean =
+                ctx.platform.classPath.findClassFile(cls.fullName.mangledString) match {
+                  case Some(entry: ZipArchive#Entry) =>
+                    entry.underlyingSource.map(_.name.startsWith("scala3-library_3-")).getOrElse(false)
```

Comment:
This assumes too much, jars can have any name and even when using a standard build tool you're not guaranteed to have the version number in the jar name (e.g. if I publish locally with sbt, the jar in ~/.ivy2/local will just be called `scala3-library_3.jar`). I would rather assume that anything in `package scala` is fine (after all, if people use `package scala` they're already doing something wrong because they get to access `private[scala]` members).

Reply:
But aren't there plenty of libraries such as the scala combinator library that are official and using the scala package but not part of the std lib

Reply:
There's also `scala.xml`, `scala.meta`, `scala.cli` and probably a few others.
Maybe then `"scala3-library"` prefix would cover the cases that we care about?